commit f2f1fa78a155524b849edf359e42a3001ea652c0
Author: Linus Torvalds <torvalds@linux-foundation.org>
Date:   Fri Dec 5 14:49:18 2008 -0800

    Enforce a minimum SG_IO timeout
    
    There's no point in having too short SG_IO timeouts, since if the
    command does end up timing out, we'll end up through the reset sequence
    that is several seconds long in order to abort the command that timed
    out.
    
    As a result, shorter timeouts than a few seconds simply do not make
    sense, as the recovery would be longer than the timeout itself.
    
    Add a BLK_MIN_SG_TIMEOUT to match the existign BLK_DEFAULT_SG_TIMEOUT.
    
    Suggested-by: Alan Cox <alan@lxorguk.ukuu.org.uk>
    Acked-by: Tejun Heo <tj@kernel.org>
    Acked-by: Jens Axboe <jens.axboe@oracle.com>
    Cc: Jeff Garzik <jeff@garzik.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Adjusted to apply to Debian's 2.6.26 by dann frazier <dannf@hp.com>

diff -urpN linux-source-2.6.26.orig/block/bsg.c linux-source-2.6.26/block/bsg.c
--- linux-source-2.6.26.orig/block/bsg.c	2009-01-08 16:43:12.000000000 -0700
+++ linux-source-2.6.26/block/bsg.c	2009-01-09 18:03:55.000000000 -0700
@@ -201,6 +201,8 @@ static int blk_fill_sgv4_hdr_rq(struct r
 		rq->timeout = q->sg_timeout;
 	if (!rq->timeout)
 		rq->timeout = BLK_DEFAULT_SG_TIMEOUT;
+	if (rq->timeout < BLK_MIN_SG_TIMEOUT)
+		rq->timeout = BLK_MIN_SG_TIMEOUT;
 
 	return 0;
 }
diff -urpN linux-source-2.6.26.orig/block/scsi_ioctl.c linux-source-2.6.26/block/scsi_ioctl.c
--- linux-source-2.6.26.orig/block/scsi_ioctl.c	2009-01-08 16:43:12.000000000 -0700
+++ linux-source-2.6.26/block/scsi_ioctl.c	2009-01-09 18:03:55.000000000 -0700
@@ -232,6 +232,8 @@ static int blk_fill_sghdr_rq(struct requ
 		rq->timeout = q->sg_timeout;
 	if (!rq->timeout)
 		rq->timeout = BLK_DEFAULT_SG_TIMEOUT;
+	if (rq->timeout < BLK_MIN_SG_TIMEOUT)
+		rq->timeout = BLK_MIN_SG_TIMEOUT;
 
 	return 0;
 }
diff -urpN linux-source-2.6.26.orig/include/linux/blkdev.h linux-source-2.6.26/include/linux/blkdev.h
--- linux-source-2.6.26.orig/include/linux/blkdev.h	2008-07-13 15:51:29.000000000 -0600
+++ linux-source-2.6.26/include/linux/blkdev.h	2009-01-09 18:03:55.000000000 -0700
@@ -585,6 +585,7 @@ extern unsigned long blk_max_low_pfn, bl
  * default timeout for SG_IO if none specified
  */
 #define BLK_DEFAULT_SG_TIMEOUT	(60 * HZ)
+#define BLK_MIN_SG_TIMEOUT	(7 * HZ)
 
 #ifdef CONFIG_BOUNCE
 extern int init_emergency_isa_pool(void);
